pep_sphinx_extensions/pep_zero_generator/writer.py

```diff
@@ -0,0 +1,316 @@
+        self.emit_title("Authors/Owners", "authors")
+        self.emit_author_table_separator(max_name_len)
+        self.emit_text(f"{'Name':{max_name_len}}  Email Address")
+        self.emit_author_table_separator(max_name_len)
+        for author_name in _sort_authors(authors_dict):
+            # Use the email from authors_dict instead of the one from "author" as
+            # the author instance may have an empty email.
+            self.emit_text(f"{author_name:{max_name_len}}  {authors_dict[author_name]}")
+        self.emit_author_table_separator(max_name_len)
+        self.emit_newline()
+        self.emit_newline()
+
+        # References for introduction footnotes
+        self.emit_title("References", "references")
+        self.emit_text(references)
+
+        pep0_string = "\n".join([str(s) for s in self.output])
+        return pep0_string
+
+
+def _classify_peps(peps: list[PEP]) -> tuple[list[PEP], ...]:
+    """Sort PEPs into meta, informational, accepted, open, finished,
+    and essentially dead."""
+    remaining = set(peps)
+
+    # The order of the comprehensions below is important. Key status values
+    # take precedence over type value, and vice-versa.
+    open_ = sorted(pep for pep in remaining if pep.status == STATUS_DRAFT)
+    remaining -= {pep for pep in open_}
+
+    deferred = sorted(pep for pep in remaining if pep.status == STATUS_DEFERRED)
+    remaining -= {pep for pep in deferred}
+
+    meta = sorted(pep for pep in remaining if pep.pep_type == TYPE_PROCESS and pep.status == STATUS_ACTIVE)
+    remaining -= {pep for pep in meta}
+
+    dead = [pep for pep in remaining if pep.pep_type == TYPE_PROCESS and pep.status in {STATUS_WITHDRAWN, STATUS_REJECTED}]
+    remaining -= {pep for pep in dead}
+
+    historical = [pep for pep in remaining if pep.pep_type == TYPE_PROCESS]
+    remaining -= {pep for pep in historical}
+
+    dead = sorted(dead + [pep for pep in remaining if pep.status in DEAD_STATUSES])
+    remaining -= {pep for pep in dead}
+
+    # Hack until the conflict between the use of `Final`
+    # for both API definition PEPs and other (actually
+    # obsolete) PEPs is addressed
+    info = sorted(
+        pep for pep in remaining
+        if pep.pep_type == TYPE_INFO and (pep.status == STATUS_ACTIVE or "Release Schedule" not in pep.title)
+    )
+    remaining -= {pep for pep in info}
+
+    historical = sorted(historical + [pep for pep in remaining if pep.pep_type == TYPE_INFO])
+    remaining -= {pep for pep in historical}
+
+    provisional = sorted(pep for pep in remaining if pep.status == STATUS_PROVISIONAL)
+    remaining -= {pep for pep in provisional}
+
+    accepted = sorted(pep for pep in remaining if pep.status in {STATUS_ACCEPTED, STATUS_ACTIVE})
+    remaining -= {pep for pep in accepted}
+
+    finished = sorted(pep for pep in remaining if pep.status == STATUS_FINAL)
+    remaining -= {pep for pep in finished}
```

Comment:
I find the iteration over remaining a bit noisy, is the whole set so big that doing the iteration over the whole set in all of them so expensive? Otherwise I would prefer to just drop `remaining` and leave the individual comprehensions.

Reply:
No, but we need to track processed PEPs somehow - the comment on line 223 `The order of the comprehensions below is important. Key status values take precedence over type value, and vice-versa.` is by Ben Peterson from 2008

Unfortunately those precedence rules are not all mutually exclusive - in testing the comprehensions vs old big for-loop, there were significant differences without tracking of some sort, and the `remaining` method I thought was cleanest, instead of `pep not in classified_peps` checks in every comprehension.

Would you have any other ideas for this algorithmically?

Reply:
> iteration over the whole set in all of them so expensive

Iteration over the whole set with some sort of filtering condition in a comprehension takes ~30-40 microseconds from memory (I did test!) so that wasn't an issue, to confirm.

Reply:
> No, but we need to track processed PEPs somehow - the comment on line 223 `The order of the comprehensions below is important. Key status values take precedence over type value, and vice-versa.` is by Ben Peterson from 2008

Ugh, I have missed that important detail.  In that case I would prefer to just preserve the origina big `if/elif`/elif` chain to reduce mistakes. Sorry about that :(

Reply:
We can land afterwards

Reply:
Sounds good, done

Reply:
Left the comprehension work in the git history too so easier to pick up & work on later